lib/documents.js

```diff
@@ -2266,9 +2267,39 @@ function writeAllDocumentsImpl(options) {
   operation.stream = inputStream;
   let result = requester.startRequest(operation);
   operation.response = result;
+  const writeBatchArray = [];
+  const batchSize = options.batchSize === null? 100:options.batchSize;
+
+  a:
+  for(let i=0; i<batchSize; i++) {
+    const record = inputStream.read();
+    if(record === null) {
+      switch(writeBatchArray.length) {
+        case 0: { finishWriter(options.writerId, done); break a;}
+        case batchSize: {
+          writeDocumentList.call(this, writeBatchArray, requestOptions);break a;}
+        default: {writeDocumentList.call(this, writeBatchArray, requestOptions).result((out) =>
+            finishWriter(writerId, done)).catch((err) => done(err));break a;}
+      }
+    }
+    writeBatchArray.push(record);
+  }
+
+  if(writeBatchArray.length !== 0) {
+    writeDocumentList.call(this, writeBatchArray, requestOptions).result((output) => writeAllDocumentsImpl(writerId, done))
+        .catch((err) => console.log(err));
+  }
+
```

Comment:
This new code must execute when it is time to write a new batch of documents and not when the initialization request starts. 

I'd suggest implementing a new onWriteAllInit() function.

Pass a reference to the new function as the callback for the initialization response, as in:

`requester.startRequest(operation).result(mlutil.callbackOn(operation, onWriteAllInit));
`

Eventually, it will be important to pass an error handler as the second argument to result(), but that can wait.

The mlutil.callbackOn() ensures that this is a reference to the operation object with the state.

Suggestions for the body of the onWriteAllInit(output) function:

```
function onWriteAllInit(output) {
    const operation = this;
   // get the number of hosts and forests from the output
   // set the number of hosts and forests in the operation
   // where the phase 2 prototype used 3 for the maximum number of requesters,
   //     instead use the number of hosts for now; and a configured multiple eventually
   for (let i=0; i < maxRequesters; i++ {
       // start the requests as in the phase 2 prototype, but pass the operation object
   }
}
```

The code added in lines 2270 through 2290 above belongs in function that's called from the for loop in onWriteAllInit() and from the callback for the writeDocumentList() result.

Use the operation object instead of this.